File: Python/Design/M-208. Implement Trie (Prefix Tree).py
class TrieNode:
    def __init__(self) -> None:
        self.children = {}
        self.is_end_char = False


class Trie:
    def __init__(self) -> None:
        self.root = TrieNode()

    def insert(self, word: str) -> None:
        node = self.root
        for char in word:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.is_end_char = True

    def is_matched(self, word: str, fully_matched: bool = True) -> bool:
        node = self.root
        for char in word:
            if char not in node.children:
                return False
            node = node.children[char]
        return node.is_end_char if fully_matched else True

    def search(self, word: str) -> bool:
        return self.is_matched(word)

    def startsWith(self, prefix: str) -> bool:
        return self.is_matched(prefix, False)


# Your Trie object will be instantiated and called as such:
# obj = Trie()
# obj.insert(word)
# param_2 = obj.search(word)
# param_3 = obj.startsWith(prefix)

File: Javascript/Design/M-208. Implement Trie (Prefix Tree).ts
class TrieNode {
  children: object;
  isEndChar: boolean;
  constructor() {
    this.children = {};
    this.isEndChar = false;
  }
}

class Trie {
  root: TrieNode;
  constructor() {
    this.root = new TrieNode();
  }
  insert(word: string): void {
    let node = this.root;
    for (const char of word) {
      if (!(char in node.children)) node.children[char] = new TrieNode();
      node = node.children[char];
    }
    node.isEndChar = true;
  }

  isMatched(word: string, fullyMatched = true): boolean {
    let node = this.root;
    for (const char of word) {
      if (!(char in node.children)) return false;
      node = node.children[char];
    }
    return fullyMatched ? node.isEndChar : true;
  }

  search(word: string): boolean {
    return this.isMatched(word);
  }

  startsWith(prefix: string): boolean {
    return this.isMatched(prefix, false);
  }
}

/**
 * Your Trie object will be instantiated and called as such:
 * var obj = new Trie()
 * obj.insert(word)
 * var param_2 = obj.search(word)
 * var param_3 = obj.startsWith(prefix)
 */
